fe1-web/network/NetworkConnection.ts

```diff
@@ -23,65 +24,83 @@ interface PendingResponse {
 }
 
 export class NetworkConnection {
-  private readonly ws: any;
+  private ws: W3CWebSocket;
 
   private payloadSentCount: number = 0;
 
   private payloadPending: Map<number, PendingResponse> = new Map();
 
   private onMessageHandler: ResponseHandler = DEFAULT_RESPONSE_HANDLER;
 
-  public readonly address: string; // FIXME get the address from ws (log ws when app is running)
+  public readonly address: string;
 
   constructor(address: string) {
-    const ws = new W3CWebSocket(address);
+    this.ws = this.establishConnection(address);
+    this.address = this.ws.url;
+  }
+
+  private establishConnection(address: string): W3CWebSocket {
+    const ws: W3CWebSocket = new W3CWebSocket(address);
 
     ws.onopen = () => { console.info(`Initiating web socket : ${address}`); };
     ws.onmessage = (message: any) => {
       console.log(`Received a new message from '${address}' : `, message.data);
 
-      let response: JsonRpcResponse;
+      let parsedMessage: JsonRpcResponse | JsonRpcRequest;
       try {
-        response = JsonRpcResponse.fromJson(message.data);
+        parsedMessage = JsonRpcResponse.fromJson(message.data);
       } catch (e) {
-        throw new ProtocolError(`Answer is not valid Json : ${e}`);
+        try {
+          parsedMessage = JsonRpcRequest.fromJson(message.data);
+        } catch (ee) {
+          throw new ProtocolError(`Answer is not valid Json : ${e} && ${ee}`);
+        }
       }
 
-      if (response.id !== UNDEFINED_ID) {
-        const pendingResponse = this.payloadPending.get(response.id);
-        if (pendingResponse === undefined) {
-          throw new NetworkError(`Received a response which id = ${response.id} does not match any pending requests`);
-        }
+      if (parsedMessage instanceof JsonRpcResponse) {
+        // if we have a response
```

Comment:
I would refactor for readability this as 
```
try {
  let parsedMesssage: JsonRpcResponse | JsonRpcRequest = this.parseWsMessage(message.data); // new method
  if (parsedMessage instanceof JsonRpcResponse) {
    this.processResponse(parsedMessage); // new method
  } else {
    this.processRequest(parsedMessage); // new method
  } 
} catch (e) {
  console.warn('Received invalid message: data.message');
}
```

Reply:
@RaulinN this answers your quesoint.
If you process it in a this.parsedWsMessage you only have to deal with the try-catch in that function.